### PR TITLE
Feature: Pretty explain in SQL Editor 

### PR DESCRIPTION
**Pull request by awaseem**
## I have read the [CONTRIBUTING.md](https://github.com/supabase/supabase/blob/master/CONTRIBUTING.md) file.

YES/NO

## What kind of change does this PR introduce?

Feature: make it easier to understand explain

## How to test

Create a couple of tables with some data

```
-- Create sample tables
CREATE TABLE users (
  id SERIAL PRIMARY KEY,
  name TEXT NOT NULL,
  email TEXT UNIQUE NOT NULL,
  age INT,
  created_at TIMESTAMP DEFAULT NOW()
);

CREATE TABLE orders (
  id SERIAL PRIMARY KEY,
  user_id INT REFERENCES users(id),
  product TEXT NOT NULL,
  amount DECIMAL(10, 2),
  status TEXT DEFAULT 'pending',
  created_at TIMESTAMP DEFAULT NOW()
);

-- Insert some sample data
INSERT INTO users (name, email, age)
SELECT 
  'User ' || i,
  'user' || i || '@example.com',
  20 + (i % 50)
FROM generate_series(1, 1000) i;

INSERT INTO orders (user_id, product, amount, status)
SELECT 
  (i % 1000) + 1,
  'Product ' || (i % 100),
  (random() * 500)::decimal(10,2),
  CASE (i % 4) 
    WHEN 0 THEN 'pending'
    WHEN 1 THEN 'shipped'
    WHEN 2 THEN 'delivered'
    ELSE 'cancelled'
  END
FROM generate_series(1, 5000) i;
```

Then use the SQL editor for some cool explain magic

```
EXPLAIN ANALYZE
SELECT u.*, o.product
FROM users u
INNER JOIN orders o ON u.id = o.user_id
WHERE o.amount > 200;

EXPLAIN ANALYZE 
SELECT u.name, COUNT(o.id) as order_count, SUM(o.amount) as total
FROM users u
LEFT JOIN orders o ON u.id = o.user_id
WHERE u.age >= 25
GROUP BY u.id, u.name
ORDER BY total DESC NULLS LAST
LIMIT 10;

EXPLAIN ANALYZE 
SELECT u.name, o.product, o.amount 
FROM users u 
JOIN orders o ON u.id = o.user_id 
WHERE o.status = 'shipped';
```

## Summary by CodeRabbit

* **New Features**
  * Interactive EXPLAIN visualization: diagram view with node rendering, icons, descriptions, nested layout, cost bars, row‑count indicators, and tooltips.
  * Header with visual/raw toggle, summary stats (total time, scan warnings) and legend; results panel can switch between diagram and raw grid.
  * Parser, types and utilities to convert EXPLAIN output into a hierarchical plan consumed by the visualizer.

* **Tests**
  * Extensive parser tests covering EXPLAIN/ANALYZE scenarios, parallel plans, nested/edge cases and metrics.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

# Demo

https://github.com/user-attachments/assets/a6c4656c-d70a-4863-b6dc-4a6e63edad7c

### CODE REVIEW

**Review by kemaldotearth — APPROVED** [review]
🚢  it's already such a step forward and gives us a platform to go off!

File: apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.CostIndicator.tsx [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+function getCostColorClass(level: CostLevel): string {
+  switch (level) {
+    case 'low':
+      return 'bg-brand'
+    case 'moderate':
+      return 'bg-warning'
+    case 'high':
+      return 'bg-destructive'
+  }
+}
```
[the remark this answers is not in the document]
Comment (awaseem) [reply apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.CostIndicator.tsx:26]: Great suggestion actually going to add

File: apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.utils.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  }
+  if (op.includes('index scan')) {
+    return 'Uses index to find matching rows'
+  }
+  if (op.includes('bitmap index scan')) {
+    return 'Builds bitmap of matching rows from index'
+  }
+  if (op.includes('bitmap heap scan')) {
+    return 'Fetches rows using bitmap'
+  }
```
[the remark this answers is not in the document]
Comment (awaseem) [reply apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.utils.ts:29]: Wow found an actual bug, fixing

File: apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.parser.ts [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    }
+    if (node.cost) {
+      stats.totalCost = Math.max(stats.totalCost, node.cost.end)
+    }
+    const op = node.operation.toLowerCase()
+    if (op.includes('seq scan')) {
+      stats.hasSeqScan = true
+      const tableMatch = node.details.match(/on\s+(\w+)/)
+      if (tableMatch) stats.seqScanTables.push(tableMatch[1])
+    }
```
[the remark this answers is not in the document]
Comment (awaseem) [reply apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.parser.ts:252]: nice another bug, pretty helpful

File: apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.CostIndicator.tsx [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+          </span>
+        </div>
+      </TooltipTrigger>
+      <TooltipContent side="top" className="max-w-xs">
+        <p className="font-medium">{label}</p>
+        <p className="text-foreground-lighter text-xs mt-1">{description}</p>
+      </TooltipContent>
+    </Tooltip>
+  )
+}
```
[the remark this answers is not in the document]
Comment (awaseem) [reply apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.CostIndicator.tsx:77]: I don't think this is needed for now

File: apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.CostIndicator.tsx [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  maxCost: number
+}
+
+type CostLevel = 'low' | 'moderate' | 'high'
+
+function getCostLevel(cost: number): CostLevel {
+  if (cost < 100) return 'low'
+  if (cost < 1000) return 'moderate'
+  return 'high'
+}
```
[the remark this answers is not in the document]
Comment (awaseem) [reply apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.CostIndicator.tsx:15]: Also fine for now

File: apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.utils.ts [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+  ListFilter,
+  SortAsc,
+  Zap,
+  Layers,
+  type LucideIcon,
+} from 'lucide-react'
+import type { ExplainNode } from './ExplainVisualizer.types'
+
+// Get human-readable description for an operation
+export function getOperationDescription(operation: string, node: ExplainNode): string {
```
Comment (awaseem) [inline apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.utils.ts]: love me some if statement soup

[the remark this answers is not in the document]
Comment (awaseem) [reply apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.utils.ts]: damn another good call out

File: apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.Header.tsx [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+      <div className="flex items-center gap-2">
+        <Activity size={16} className="text-foreground-light" />
+        <h3 className="text-sm font-medium text-foreground">Query Execution Plan</h3>
+        <button
+          type="button"
+          onClick={onToggleMode}
+          className="font-mono text-xs text-foreground-lighter hover:text-foreground transition-colors"
+          aria-label={isVisual ? 'Switch to raw explain output' : 'Switch to visual explain output'}
+        >
+          {isVisual ? '[RAW]' : '[VISUAL]'}
```
Comment (charislam) [inline apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.Header.tsx]: I actually find this intuitively backwards for me. It looks to me more like a label than a button, so it seems to be announcing that this is the "RAW" view, rather than that you should click the button to get to the "RAW" view, if that makes any sense

Image

File: apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.NodeRenderer.tsx [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    <div className="relative">
+      {/* Node content */}
+      <div
+        className={cn(
+          'flex items-start gap-3 py-3 px-3 rounded-lg border border-transparent hover:border-border hover:bg-surface-100 transition-all'
+        )}
+        style={{ marginLeft: `${depth * 32}px` }}
+      >
+        {/* Pipeline flow indicator */}
+        <div className="flex flex-col items-center gap-1">
```
Comment (charislam) [inline apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.NodeRenderer.tsx]: unnecessary div?

File: apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.NodeRenderer.tsx [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+} from './ExplainVisualizer.utils'
+
+import type { ExplainNode } from './ExplainVisualizer.types'
+
+export interface ExplainNodeRendererProps {
+  node: ExplainNode
+  isLast: boolean
+  depth: number
+  maxCost: number
+  parentHasMore: boolean[]
```
Comment (charislam) [inline apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.NodeRenderer.tsx]: are `parentMore` and `isLast` actually used? I see them passed through the recursion but couldn't find anywhere in the component that actually uses them

File: apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.parser.ts [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+    level,
+    children: [],
+    raw,
+  }
+
+  if (metrics) {
+    // Parse cost=start..end
+    const costMatch = metrics.match(/cost=([\d.]+)\.\.([\d.]+)/)
+    if (costMatch) {
+      node.cost = { start: parseFloat(costMatch[1]), end: parseFloat(costMatch[2]) }
```
Comment (charislam) [inline apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.parser.ts]: should NaN checks be done here (for all metrics)? instead of moving defensiveness further down, we can guarantee that everything is either a number or undefined

File: apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.RowCountIndicator.tsx [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+    if (ratio < 0.5) return 'overestimated'
+    return 'accurate'
+  }
+
+  if (showFilterFlow) {
+    return (
+      <Tooltip>
+        <TooltipTrigger asChild>
+          <div className="flex items-center gap-1.5 text-xs text-foreground-lighter cursor-help">
+            <span>{totalRowsScanned!.toLocaleString()} rows</span>
```
Comment (charislam) [inline apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.RowCountIndicator.tsx]: minor nit, but would be nice to have a utility function that returns the display branch with type safety, e.g., something that returns:

`{type: 'filter', totalRowsScanned: number} | { type: 'simple' ...`

File: apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.RowCountIndicator.tsx [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+            The planner's estimate matched the actual row count exactly.
+          </p>
+        </TooltipContent>
+      </Tooltip>
+    )
+  }
+
+  if (showComparison) {
+    const accuracy = getEstimationAccuracy()
+    const hasEstimationIssue = accuracy && !['accurate'].includes(accuracy)
```
Comment (charislam) [inline apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.RowCountIndicator.tsx]:
```suggestion
    const hasEstimationIssue = accuracy && accuracy !== 'accurate'
```

unless it's likely that the exception cases will expand later?

File: apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.RowCountIndicator.tsx [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+                  <ArrowRight size={10} className="text-foreground-muted" />
+                )}
+                <span className="text-foreground-light font-medium">
+                  {actualRows.toLocaleString()} rows
+                </span>
+              </>
+            )}
+          </div>
+        </TooltipTrigger>
+        <TooltipContent side="top" className="max-w-xs">
```
Comment (charislam) [inline apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.RowCountIndicator.tsx]: There's a lot of conditional logic in different ternaries here, might be clearer if we split these out into different subconditions

File: apps/studio/tests/features/explain-visualizer/ExplainVisualizer.parser.test.ts [hunk 14]
```diff
@@ -0,0 +1,1100 @@
...
+        ],
+        raw: '',
+      },
+    ]
+
+    const result = calculateSummary(tree)
+    expect(result.hasIndexScan).toBe(true)
+  })
+
+  test('finds maximum cost across nested nodes', () => {
```
Comment (charislam) [inline apps/studio/tests/features/explain-visualizer/ExplainVisualizer.parser.test.ts]: does this test the same thing as `finds maximum across nested children`?

**Review by charislam — APPROVED** [review]

File: apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.Header.tsx (hunk 7 above)
Comment (awaseem) [reply apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.Header.tsx]: Oh yeah, that makes way more sense

File: apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.NodeRenderer.tsx (hunk 8 above)
Comment (awaseem) [reply apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.NodeRenderer.tsx]: Damn yes! AI slop on my part

File: apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.NodeRenderer.tsx (hunk 9 above)
Comment (awaseem) [reply apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.NodeRenderer.tsx]: 100% right on this, they were left over from an older design. Nice catch

File: apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.RowCountIndicator.tsx (hunk 12 above)
Comment (awaseem) [reply apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.RowCountIndicator.tsx]: what the, also AI slop will fix

File: apps/studio/tests/features/explain-visualizer/ExplainVisualizer.parser.test.ts (hunk 14 above)
Comment (awaseem) [reply apps/studio/tests/features/explain-visualizer/ExplainVisualizer.parser.test.ts]: Yeah it is, the other one is checking something totally useless so removing

File: apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.RowCountIndicator.tsx (hunk 13 above)
Comment (awaseem) [reply apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.RowCountIndicator.tsx]: Yeah totally, hard to read

File: apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.parser.ts (hunk 10 above)
Comment (awaseem) [reply apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.parser.ts]: 100000% yes cleaned that all up

File: apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.RowCountIndicator.tsx (hunk 11 above)
Comment (awaseem) [reply apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.RowCountIndicator.tsx]: I split the functions a bit, should be easier to read and process

File: apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.RowCountIndicator.tsx [hunk 15]
```diff
@@ -2,138 +2,183 @@ import { ArrowRight } from 'lucide-react'
...
+const formatCount = (value?: number): string => value?.toLocaleString() ?? 'Unknown'
+
+type EstimationAccuracy =
+  | 'severely underestimated'
+  | 'underestimated'
+  | 'severely overestimated'
+  | 'overestimated'
+  | 'accurate'
+
+function getEstimationAccuracy(actualRows: number, estimatedRows: number): EstimationAccuracy {
```
Comment (charislam) [inline apps/studio/components/interfaces/ExplainVisualizer/ExplainVisualizer.RowCountIndicator.tsx:20]: guard against a divide by 0?

File: apps/studio/tests/features/explain-visualizer/ExplainVisualizer.parser.test.ts [hunk 16]
```diff
@@ -517,6 +517,95 @@ describe('parseExplainOutput', () => {
...
+
+      const result = parseExplainOutput(input)
+      parseNodeDetails(result[0])
+
+      expect(result).toHaveLength(1)
+      // Should not parse invalid value at all (regex won't match)
+      expect(result[0].rowsRemovedByFilter).toBeUndefined()
+    })
+
+    test('ensures all numeric values are finite after parsing', () => {
```
Comment (charislam) [inline apps/studio/tests/features/explain-visualizer/ExplainVisualizer.parser.test.ts]: Nit: claim is a bit broader than the test ensures, because the metrics are all explicitly defined, so this is more checking, e.g., that `node.cost.start` is `0`, and we know 0 to be finite, than checking that it will never produce an infinite value (that would require a property-based test)

**Review by charislam — APPROVED** [review]

File: apps/studio/tests/features/explain-visualizer/ExplainVisualizer.parser.test.ts (hunk 16 above)
Comment (awaseem) [reply apps/studio/tests/features/explain-visualizer/ExplainVisualizer.parser.test.ts]: I just updated the description to be more granular
